packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,364 @@
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { AddressResolver } from "../common/AddressResolver.sol";
+import { EssentialContract } from "../common/EssentialContract.sol";
+import { IProverPool } from "./IProverPool.sol";
+import { LibMath } from "../libs/LibMath.sol";
+import { TaikoToken } from "./TaikoToken.sol";
+import { Proxied } from "../common/Proxied.sol";
+
+contract ProverPool is EssentialContract, IProverPool {
+    using LibMath for uint256;
+    // 8 bytes or 1 uint64
+
+    struct Prover {
+        uint64 stakedAmount;
+        uint16 rewardPerGas;
+        uint16 currentCapacity;
+        uint64 weight;
+    }
+
+    // Make sure we only use one slot
+    struct Staker {
+        uint64 exitRequestedAt;
+        uint64 exitAmount;
+        uint16 maxCapacity;
+        uint8 proverId; // 0 to indicate the staker is not a top prover
+    }
+
+    // Given that we only have 32 slots for the top provers, if the protocol
+    // can support 1 block per second with an average proof time of 1 hour,
+    // then we need a min capacity of 3600, which means each prover shall
+    // provide a capacity of at least 3600/32=112.
+    uint32 public constant MAX_CAPACITY_LOWER_BOUND = 128;
```

Comment:
In A4, I think the proof won't cost 1 hour so the lower bound may be too big.

Reply:
Though, it's not big issue.

Reply:
We assume eventually each prove may take ~1 hour.